### PR TITLE
Add CHAP support

### PR DESCRIPTION
**Pull request by NaveenSreeramachandra**
Following changes were made to support CHAP:

1) Support for 1-way and 2-way CHAP through StorageClass.
2) Configure both Datera Storage Instance Auth and ISCSI CHAP on K8S worker nodes.
3) Hide CHAP credentials in all GRPC and CSI driver logs messages.
4) Add an example file showing how to setup CHAP and update README.
5) Provide new CSI driver installation YAML version 1.0.10.

### CODE REVIEW

File: pkg/common/util.go [hunk 1]
```diff
@@ -160,3 +161,47 @@ func versionToInt(v string) (int, error) {
 	}
 	return result, nil
 }
+
+func StripSecretsAndGetChapParams(req interface{}) (map[string]string) {
+
+	strip_needed := false
```
Comment (virgilm) [inline pkg/common/util.go:167]: use camelCase

File: pkg/common/util.go [hunk 2]
```diff
@@ -160,3 +161,47 @@ func versionToInt(v string) (int, error) {
...
+	chap_params := map[string]string{}
```
Comment (virgilm) [inline pkg/common/util.go:168]: same here, use chapParams instead

File: pkg/common/util.go [hunk 3]
```diff
@@ -160,3 +161,47 @@ func versionToInt(v string) (int, error) {
...
+		case *csi.CreateVolumeRequest:
+			secrets = reqType.Secrets
+			strip_needed = true
+		case *csi.NodeStageVolumeRequest:
+                        secrets = reqType.Secrets
+                        strip_needed = true
+		case *csi.DeleteVolumeRequest:
+			secrets = reqType.Secrets
+                        strip_needed = true
+                default:
```
Comment (virgilm) [inline pkg/common/util.go:181]: align

File: pkg/driver/controller.go [hunk 4]
```diff
@@ -199,6 +199,10 @@ func registerMdFromCtxt(ctxt context.Context, md *dc.VolMetadata) error {
 }
 
 func (d *Driver) CreateVolume(ctx context.Context, req *csi.CreateVolumeRequest) (*csi.CreateVolumeResponse, error) {
+
+	chap_params := map[string]string{}
```
Comment (virgilm) [inline pkg/driver/controller.go:203]: use chapParams

File: pkg/driver/node.go [hunk 5]
```diff
@@ -16,6 +16,10 @@ import (
 )
 
 func (d *Driver) NodeStageVolume(ctx context.Context, req *csi.NodeStageVolumeRequest) (*csi.NodeStageVolumeResponse, error) {
+
+        chap_params := map[string]string{}
+        chap_params = co.StripSecretsAndGetChapParams(req)
```
Comment (virgilm) [inline pkg/driver/node.go:21]: use chapParams instead

File: pkg/driver/controller.go [hunk 6]
```diff
@@ -330,6 +339,8 @@ func (d *Driver) CreateVolume(ctx context.Context, req *csi.CreateVolumeRequest)
 }
 
 func (d *Driver) DeleteVolume(ctx context.Context, req *csi.DeleteVolumeRequest) (*csi.DeleteVolumeResponse, error) {
+
+	_ = co.StripSecretsAndGetChapParams(req)
```
Comment (virgilm) [inline pkg/driver/controller.go:343]: hmm, is this doing anything?

Comment (NaveenSreeramachandra) [reply pkg/driver/controller.go:343]: Yes, stripping the secrets only. The chap parameters returned are discarded. It is not needed for DeleteVolume case.

File: pkg/client/login.go [hunk 7]
```diff
@@ -35,14 +35,51 @@ func (v *Volume) Login(multipath, round_robin bool) error {
...
+		c.AuthType = "chap"
+		c.SessionSecrets = secrets
+		c.DoCHAPDiscovery = true
+
+	} else {
+		c.DoDiscovery = true
+	}
+
+	for _, t := range c.Targets {
+		co.Debugf(ctxt, t.Portal)
```
[the remark this answers is not in the document]
Comment (NaveenSreeramachandra) [reply pkg/client/login.go:74]: That was a debug enabled during coding. Have removed the debug statement and the for-loop.

File: pkg/driver/node.go (hunk 5 above)
Comment (NaveenSreeramachandra) [reply pkg/driver/node.go]: Done.

File: pkg/driver/controller.go (hunk 4 above)
Comment (NaveenSreeramachandra) [reply pkg/driver/controller.go]: Done.

File: pkg/common/util.go (hunk 3 above)
Comment (NaveenSreeramachandra) [reply pkg/common/util.go]: Done.

File: pkg/common/util.go (hunk 2 above)
Comment (NaveenSreeramachandra) [reply pkg/common/util.go]: Done

File: pkg/client/login.go [hunk 8]
```diff
@@ -35,14 +35,51 @@ func (v *Volume) Login(multipath, round_robin bool) error {
...
+        c.Multipath = multipath
+	c.RetryCount = 3
+
+	if len(chap_params) != 0 {
+		secrets.SecretsType = "chap"
+		if value, exists := chap_params["node.session.auth.username"]; exists {
+			secrets.UserName = value
+		}
+		if value, exists := chap_params["node.session.auth.password"]; exists {
+                        secrets.Password = value
```
[the remark this answers is not in the document]
Comment (NaveenSreeramachandra) [reply pkg/client/login.go]: Done.

File: pkg/client/login.go [hunk 9]
```diff
@@ -35,14 +35,51 @@ func (v *Volume) Login(multipath, round_robin bool) error {
...
+	var targets []iscsi.TargetInfo
+	for _, Ip := range ips {
+		targets = append(targets, iscsi.TargetInfo{v.Iqn, Ip, "3260"})
 	}
-	co.Debugf(ctxt, "ISCSI Connector: %#v", c)
+
+	secrets := iscsi.Secrets{}
+	c := iscsi.Connector{}
+
+        c.Targets = targets
```
[the remark this answers is not in the document]
Comment (NaveenSreeramachandra) [reply pkg/client/login.go]: Done.

File: go.sum [hunk 10]
```diff
@@ -3,18 +3,38 @@ cloud.google.com/go v0.26.0/go.mod h1:aQUYkXzVsufM+DwF1aE+0xfcU+56JwCaLick0ClmMT
...
 github.com/Datera/go-sdk v1.1.14/go.mod h1:Y9vguQE6UvgtDGnd3eyOkYpliNnhIkrZ8Ub9ACtzzgA=
+github.com/Datera/go-sdk v1.1.15 h1:uvSeZ1PaNswxhTaN/7ULyzeuqMTjtIejFqWnR+H69Fg=
+github.com/Datera/go-sdk v1.1.16-0.20200525230021-f4729081244e h1:VjyLeS7cY6GmK1nPi0b4Amr6anwCE4n2SJUqKAs3aSg=
+github.com/Datera/go-sdk v1.1.16-0.20200525230021-f4729081244e/go.mod h1:HOlJiD8eWPbKasWqvDVLM6gPBpCSpYxX/TafZgDfzX4=
+github.com/Datera/go-sdk v1.1.16-0.20200525231803-cbe7a5c8602e h1:qmjq05eTZFeVxcuAX7oTxFqla+01DYLxjBMwQWSA/4w=
+github.com/Datera/go-sdk v1.1.16-0.20200525231803-cbe7a5c8602e/go.mod h1:HOlJiD8eWPbKasWqvDVLM6gPBpCSpYxX/TafZgDfzX4=
+github.com/Datera/go-sdk v1.1.16-0.20200526014332-ec041e3544b8 h1:Sv407Y9MwGnqgUF9G6arGe52g1WcDfDeXU0z2bTF2Ys=
+github.com/Datera/go-sdk v1.1.16-0.20200526014332-ec041e3544b8/go.mod h1:HOlJiD8eWPbKasWqvDVLM6gPBpCSpYxX/TafZgDfzX4=
+github.com/Datera/go-sdk v1.1.16-0.20200526020143-b960e22a3424 h1:zgI2R3fxjIelGf0xnZ8ooTgO/wilIY/QWUYhT2oTnpU=
+github.com/Datera/go-sdk v1.1.16-0.20200526020143-b960e22a3424/go.mod h1:HOlJiD8eWPbKasWqvDVLM6gPBpCSpYxX/TafZgDfzX4=
```
[the remark this answers is not in the document]
Comment (NaveenSreeramachandra) [reply go.sum]: Will push the updated go.sum in the next version commit.

File: go.mod [hunk 11]
```diff
@@ -3,8 +3,8 @@ module github.com/Datera/datera-csi
 go 1.12
 
 require (
-	cloud.google.com/go v0.37.1 // indirect
-	github.com/Datera/go-sdk v1.1.14
+	git.apache.org/thrift.git v0.12.0 // indirect
+	github.com/Datera/go-sdk v1.1.16-0.20200526020143-b960e22a3424
```
[the remark this answers is not in the document]
Comment (NaveenSreeramachandra) [reply go.mod]: Have updated go-sdk with v1.1.16 tag and using that version in CSI driver go.mod

File: pkg/client/volume.go [hunk 12]
```diff
@@ -333,33 +333,27 @@ func (r *DateraClient) CreateVolume(name string, volOpts *VolOpts, qos bool, cha
...
-					InitiatorUserName: chap_params["node.session.auth.username_in"],
-					InitiatorPassword: chap_params["node.session.auth.password_in"],
-					TargetUserName: chap_params["node.session.auth.username"],
-					TargetPassword: chap_params["node.session.auth.password"],
-				},
-				Volumes: []*dsdk.Volume{vol},
+		// Add CHAP credentials to the Storage Instance struct
+		si.Auth = &dsdk.Auth{Type: "none"}
+
+		if username_in, exists := chapParams["node.session.auth.username_in"]; exists {
```
Comment (virgilm) [inline pkg/client/volume.go:339]: camelCase

Comment (NaveenSreeramachandra) [reply pkg/client/volume.go]: Took care of it.

**Pull request merged**